packages/workspace/src/browser/workspace-frontend-contribution.ts

```diff
@@ -418,14 +421,36 @@ export class WorkspaceFrontendContribution implements CommandContribution, Keybi
         } while (selected && exist && !overwrite);
         if (selected) {
             try {
-                await this.commandRegistry.executeCommand(CommonCommands.SAVE.id);
-                await this.fileService.copy(uri, selected, { overwrite });
+                await this.copySave(uri, selected, overwrite);
             } catch (e) {
                 console.warn(e);
             }
         }
     }
 
+    private async copySave(source: URI, selected: URI, overwrite: boolean): Promise<void> {
+        const sourceWidget = this.applicationShell.currentWidget;
```

Comment:
I didn't understand the relationship between the `source` parameter and the `sourceWidget` ref?

Should you rather try to find the widget that has `source` as URI instead?

Reply:
So the relationship is quite implicit. We can use the `currentWidget` here savely for multiple reasons. First off, we know that the `currentWidget` is definitely a saveable item by checking for it in the `isEnabled` flag of the command. Secondly, the `source` parameter is provided by the `SelectionService` which contains the currently selected editor (which is a child of the `currentWidget` afaik). 

Also, saving always uses the `currentWidget` of the `ApplicationShell` as shown in [here](https://github.com/Darkgaja/theia/blob/9333761db1e9822d63888893af82ed63dc164bf3/packages/core/src/browser/shell/application-shell.ts#L1731-L1736). Furthermore, I think there isn't even a trivial method to find a widget by its uri.

Reply:
Thanks for the explanation. Essentially we use the selection service just to get a URI and use the `currentWidget` for everything else, I really find that confusing... Since I'm the one having issues with that logic, I made a commit on top of your PR, please tell me what you think and feel free to apply those changes here: https://github.com/Darkgaja/theia/compare/save-as-create-file...marechal-p:save-as-create-file-2

My commit simplifies the command handler for `SAVE_AS`, makes most checks on the `currentWidget` before calling the `saveAs` logic, and uses said widget as source-of-truth all the way down.

Reply:
Thank you, your code is definitely easier to follow. I merged your contribution.